### PR TITLE
ddl: support drop table for label rule

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Closes https://github.com/pingcap/tidb/issues/26693. It can be reviewed after https://github.com/pingcap/tidb/pull/26511 is merged. 

### What is changed and how it works?

Add a rule for a table, show it from PD API:
```
mysql> alter table t1 attributes="attr";

# curl http://127.0.0.1:2379/pd/api/v1/config/region-label/rules
[
  {
    "id": "schema/test/t1",
    "labels": [
      {
        "key": "attr",
        "value": "true"
      },
      {
        "key": "db",
        "value": "test"
      },
      {
        "key": "table",
        "value": "t1"
      }
    ],
    "rule_type": "key-range",
    "rule": {
      "start_key": "7480000000000000ff355f720000000000fa",
      "end_key": "7480000000000000ff365f720000000000fa"
    }
  }
]
```
After GC finish:
```
# curl http://127.0.0.1:2379/pd/api/v1/config/region-label/rules
[]
```

What's Changed:
1. add GC logic for label rules
2. add `UpdateLabelRules` to batch update/delete the rules
 

### Check List 

Tests 

- [x] Unit test

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [x] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 
```release-note
None
```

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -6014,3 +6017,42 @@ func (d *ddl) AlterTableAlterPartition(ctx sessionctx.Context, ident ast.Ident,
...
+		return errors.Trace(err)
+	}
+	meta := tb.Meta()
+
+	rule := label.NewRule()
+	err = rule.ApplyAttributesSpec(spec.AttributesSpec)
+	if err != nil {
+		var sb strings.Builder
+		restoreCtx := format.NewRestoreCtx(format.RestoreStringSingleQuotes|format.RestoreKeyWordLowercase|format.RestoreNameBackQuotes, &sb)
+
```
Comment (JmPotato) [inline ddl/ddl_api.go:6033]:
```suggestion
```

File: ddl/delete_range.go [hunk 2]
```diff
@@ -272,7 +272,8 @@ func insertJobIntoDeleteRangeTable(ctx context.Context, sctx sessionctx.Context,
 		// The startKey here is for compatibility with previous versions, old version did not endKey so don't have to deal with.
 		var startKey kv.Key
 		var physicalTableIDs []int64
-		if err := job.DecodeArgs(&startKey, &physicalTableIDs); err != nil {
+		var ruleIDs []string
```
Comment (JmPotato) [inline ddl/delete_range.go:275]: Is this will be used in the later PRs?

File: ddl/label/rule.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+type RulePatch struct {
+	SetRules    []*Rule  `json:"sets"`
+	DeleteRules []string `json:"deletes"`
+}
+
+// NewRulePatch returns a patch of rules which need to be set or deleted.
+func NewRulePatch(setRules []*Rule, deleteRules []string) *RulePatch {
+	return &RulePatch{
+		SetRules:    setRules,
+		DeleteRules: deleteRules,
```
Comment (JmPotato) [inline ddl/label/rule.go:111]: When did these two slices will be updated? I didn't see any related code in this PR.

**Review by Yisaer — COMMENTED** [review]
Rest lgtm

File: store/gcworker/gc_worker.go [hunk 4]
```diff
@@ -1898,6 +1907,52 @@ func (w *GCWorker) doGCPlacementRules(dr util.DelRangeTask) (pid int64, err erro
...
+			physicalTableIDs []int64
+		)
+		if err = historyJob.DecodeArgs(&startKey, &physicalTableIDs, &ruleIDs); err != nil {
+			return
+		}
+	}
+
+	patch := label.NewRulePatch(nil, ruleIDs)
+	err = infosync.UpdateLabelRules(context.TODO(), patch)
+	return
```
[the remark this answers is not in the document]
Comment (rleungx) [reply store/gcworker/gc_worker.go:1953]: The truncate table part will be added in the later PR.

**Review by Yisaer — APPROVED** [review]

**Review by xhebox — APPROVED** [review]

**Comment by morgo** [comment]
LGTM

**Pull request merged**